src/coreclr/gc/gc.cpp

```diff
@@ -3850,6 +3850,39 @@ void region_allocator::delete_region (uint8_t* region_start)
 
     leave_spin_lock();
 }
+
+bool region_allocator::should_delete_region (uint8_t* start, int num_free_regions, int num_free_large_regions)
+{
+    size_t address_space_available = global_region_end - global_region_start;
+#ifdef MULTIPLE_HEAPS
+    size_t address_space_available_per_heap = address_space_available / gc_heap::n_heaps;
+#else //MULTIPLE_HEAPS
+    size_t address_space_available_per_heap = address_space_available;
+#endif //MULTIPLE_HEAPS
+
+    size_t address_space_in_free_regions = (num_free_regions       * region_alignment      ) +
+                                           (num_free_large_regions * large_region_alignment);
+
+    // don't keep more than a quarter of the address space in the per-heap free lists
+    if (address_space_in_free_regions >= address_space_available_per_heap / 4)
+    {
+        return true;
+    }
+
+    // get stingy if the load exceeds 50%
+    if (get_va_memory_load() > 50)
```

Comment:
should this be the first check?

Reply:
All the tests are cheap, so it probably does not matter.